web-console/knikubevirt/snapshots/snapshots.md

```diff
@@ -0,0 +1,76 @@
+# Virtual Machines Snapshots
+
+Snapshots allow users to create a copy of the VM disks and memory, mainly for restore and backup purposes. The snapshots are taken live or offline depending on if the VM is running or not. 
+This document describes the snapshot design flow.  
+
+
+
+
+## From VM List 
+
+![Offline VM list - snapshot action from a single VM kebab](img/1-00.jpg)
+When clicking on the kebab menu in the VM list, the user will be able to take a snapshot of the VM, meaning he will copy the disks, memory content, and metadata for restoring, backup, creating templates and cloning VMs.
+
+![Offline Snapshot modal - default](img/1-01.jpg)
+In the basic modal, the snapshot will take all current configuration and selected parameters.
+The user must add a descriptive name.
+![Offline Snapshot modal - selecting disks](img/1-02.jpg)
+In the disks field, the user will be able to decide if to include all disks or just some of them and if to include memory as part of the snapshot or not.
+![Offline Snapshot modal - disks customization](img/1-03.jpg)
+![Offline Snapshot modal - disks customized](img/1-04.jpg)
+
+![Offline Snapshotting status indication](img/1-05.jpg)
+![Offline Snapshotting status indication on click and kebab mode](img/1-06.jpg)
+Any virtual machine that has a snapshot in progress will show this in the state:
+When in this progress mode all actions would be disabled and the “Stop snapshot” would be available.
+
+![Offline Restoring from snapshot status](img/1-07.jpg)
+Any virtual machine that is undergoing a restoration process will show a "Restoring" indication on the status, with more details and a "Stop restoring" action on a popover. 
+
+
+## VM page - snapshots tab
+
+![Offline Shanpshot tab](img/2-00.jpg)
+The snapshot list is located under the VM. A tab item in the horizontal navigation will be dedicated for snapshots. The snapshot list is ordered by  date as a default, making the "Scheduled" snapshots appear firts.
```

Comment:
I think sorting by Name by default (as shown in the image) might A) be more consistent with the rest of OpenShift and B) make it easier to find all of the snapshots (and scheduled snapshots) for a VM side-by-side. I imagine snapshot names would be something like `vm-name-20190131-1400` or something, so that grouping would happen naturally.

If you agree, these two sentences could be tweaked to match the image.

Reply:
@andybraren Yes, consistency should always be in the top of our priorities, so I should update the text.
But, this is why I didn't want to place "scheduled" and "Taken" on the same list. spreading the scheduled snapshots across the list and not locating them on the top just makes them "invisible". Now the list doesn't look like it contains two different entities.
While placing both "scheduled" and "Taken" on the same list looks neater, it makes less sense, at least to me. I'm adding both screenshots.
I won't hold this PR for this so I'll let u take the decision if that's ok with you:
![2-02b](https://user-images.githubusercontent.com/48942365/59836270-dcf31e80-9353-11e9-8fcf-da0ef8da3eaa.jpg)
![2-02c](https://user-images.githubusercontent.com/48942365/59836272-dcf31e80-9353-11e9-836b-39f3c8dc7dd7.jpg)
Thanks

Reply:
> I did miss the "Sort by name" when I said that combined list looks better. Now I'm back to square one: Mixing Scheduled and Taken snapshot on the same list and not locating them on the top just makes them "invisible", and break the logic groups that held the combined list together. Now the list doesn't look like it contains two different entities.

Sorting by name should keep the two entities grouped if named consistently - scheduled ones would just be at the bottom of the list instead of the top if sorted alphabetically. The numbers at the end of the example name in my comment  above were intended to be a timestamp.

Ignore the (B) in my previous comment. For some reason I forgot that this list is just showing one VM's snapshot history - not all snapshots across all VMs.

Here's what more realistic names and dates might look like, sorted by name:

|Name (sort)|Date|
|---|---|
|vm-name-20190131-1300|Jan 31, 1:00 pm|
|vm-name-20190201-1300|Feb 1, 1:00 pm|
|vm-name-20190202-1300|Feb 2, 1:00 pm|
|vm-name-20190203-1300|Feb 3, 1:00 pm (scheduled)|

I agree that either reverse-sorting by name or sorting by date to put the more recent snapshots (and future scheduled ones) at the top would be ideal. I don't think I've seen that in the console before, but @openshift/team-ux-review maybe this is one table where an exception makes sense. If we think of snapshots as "backups", seeing the most recent backups first without scrolling feels right.

FWIW, one assumption that makes me more comfortable with a combined table in general is that it's probably not very common to have more than 1 or 2 scheduled snapshots. Daily snapshot "backups" make sense. Weekly snapshot "backups" make sense. But 3+ scheduled backups all with different timing strategies seems like overkill, at least to me. That assumption might be worth verifying with developers though.

> I won't hold this PR for this so I'll let u take the decision if that's ok with you

The nice thing about the PR process is that merging isn't any one person's decision 🙂. Discussions like these help the entire group stay aware of console constraints that we might consider changing in the future, and make our designs better.

This combined table design, in my opinion, is a worthwhile compromise that avoids having to introduce a new pattern or have two tables (one of which might be very small) in one page. Let's see what others think.

Reply:
> FWIW, one assumption that makes me more comfortable with a combined table in general is that it's probably not very common to have more than 1 or 2 scheduled snapshots. Daily snapshot "backups" make sense. Weekly snapshot "backups" make sense. But 3+ scheduled backups all with different timing strategies seems like overkill, at least to me. That assumption might be worth verifying with developers though.

Not sure I follow with this one, If you have just one scheduled snapshot, and that produces 3 backups max, that scheduled snapshot item will be at the bottom, creating the same case I wish to avoid. But the Name combined with the time stamp does solve it, as long as it shows the 'Next' scheduled :)

Good one :)

Reply:
I think it makes sense to stick with the flat list for now. Assuming these lists will be fairly short it should be easy for the user to scan the list and see what is scheduled. Let's create an issue to investigate further though so we don't block this PR. 
It seems like it's worth some research to better understand the length of these lists. WDYT @andybraren @Ranelim ?

Reply:
@Ranelim I forgot that names are entered manually. I'm not sure that's compatible with auto-recurring scheduled snapshots, which would mean the modal form would need to be reworked to ask whether this is a recurring/scheduled snapshot or not.

@matthewcarleton Length is one thing, but I'm a little concerned that we don't have a deep enough understanding of the technical aspect of scheduled/recurring snapshots to design something implementable here. Are they a separate object internally? Could we represent them as just the next "future" snapshot? Should they be automatically-named? Can we define a custom naming scheme?

I'm fine with the snapshot portion of this PR. The futuristic "Scheduled" portion should probably either be labeled clearly as "Future/WIP" here to be more clear, or removed from this PR and investigated in more depth later. I think that's your call @matthewcarleton, and depends on what the goal of this design repo is more generally.

Reply:
@andybraren ya it's a tricky balance of not overloading the scope of these PRs. They were originally intended to just pull in already approved designs and have them vetted in the context of OS (correct me if I'm wrong @beanh66 @lizsurette). If we are going to reconsider large portions of the flow then we should step back and document the requirements and revisit the design. I'd like to avoid scenarios where we are questioning the design in the comments, it gets very hard to track that way, and the person who has to merge this will likely not feel entirely confident that it is done. 
I think we can have a larger discussion about this and figure out the best way forward though.

Reply:
> They were originally intended to just pull in already approved designs and have them vetted in the context of OS (correct me if I'm wrong @beanh66 @lizsurette).

+1 to this. I think it depends a little bit on the context, but especially in this case I think it's fair for the KubeVirt designers to work together to come up with a good approach around snapshots and then share with the OpenShift team when they come to a nice first proposal. If all of the initial design and discussion is done here it will potentially create a "too many cooks in the kitchen" scenario :)

Reply:
I discussed this with Tomas. Quoting him here:
"uff, that is one of those things every customer has different 🙂 Some have 1-2 snapshots per VM (e.g. create a snapshot, try to change something in the VM and than if worked, delete the snapshot, if it did not, rollback from it)
and than some customers are using it as a backup and can have 100..."

So, we have no common amount of snapshots taken or scheduled. We'll just have to provide a design that looks reasonable for any number :)

Reply:
+1 I think that looks great @Ranelim 
I agree with @andybraren that we could create an issue to track some of the further work around Snapshots as objects and how that affects the ui. Can you create that and loop @andybraren in to ensure it covers his concerns?

Reply:
LGTM, I think the filter helps quite a bit, and thank you for the Google example. Long-term I think separate lists makes sense, but short-term I'm okay with this solution. And creating an issue to track the the object/technical concern sounds great. 👍